bin/kmstool-enclave-cli/main.c

```diff
@@ -219,29 +337,61 @@ static int decrypt(struct app_ctx *app_ctx, struct aws_byte_buf *ciphertext_decr
     struct aws_byte_buf ciphertext_decrypted;
     rc = aws_kms_decrypt_blocking(
         client, app_ctx->key_id, app_ctx->encryption_algorithm, &ciphertext, &ciphertext_decrypted);
-
+    
     aws_byte_buf_clean_up(&ciphertext);
     fail_on(rc != AWS_OP_SUCCESS, "Could not decrypt ciphertext");
 
     /* Encode ciphertext into base64 for printing out the result. */
-    size_t ciphertext_decrypted_b64_len;
-    struct aws_byte_cursor ciphertext_decrypted_cursor = aws_byte_cursor_from_buf(&ciphertext_decrypted);
-    aws_base64_compute_encoded_len(ciphertext_decrypted.len, &ciphertext_decrypted_b64_len);
-    rc = aws_byte_buf_init(ciphertext_decrypted_b64, app_ctx->allocator, ciphertext_decrypted_b64_len + 1);
-    fail_on(rc != AWS_OP_SUCCESS, "Memory allocation error");
-    rc = aws_base64_encode(&ciphertext_decrypted_cursor, ciphertext_decrypted_b64);
-    fail_on(rc != AWS_OP_SUCCESS, "Base64 encoding error");
-    aws_byte_buf_append_null_terminator(ciphertext_decrypted_b64);
+    rc = encode_b64(app_ctx, &ciphertext_decrypted, ciphertext_decrypted_b64);
+    fail_on(rc != AWS_OP_SUCCESS, "Could not encode ciphertext");
 
+    /* Cleaning up allocated memory */
     aws_nitro_enclaves_kms_client_destroy(client);
     aws_credentials_release(credentials);
+    
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * Function to generate a data key from KMS with attestation.
+ *
+ * @param[in]  app_ctx: Struct that has all of the necessary arguments
+ * @param[out] ciphertext_decrypted_b64: Byte buffer where the ciphertext blob will be stored
+ * @param[out] plaintext_b64: Byte buffer where the plaintext output will be stored
+ */
+static int gen_datakey(struct app_ctx *app_ctx, struct aws_byte_buf *ciphertext_b64, struct aws_byte_buf *plaintext_b64) {
+    ssize_t rc = 0;
+
+    struct aws_credentials *credentials = NULL;
+    struct aws_nitro_enclaves_kms_client *client = NULL;
+
+    init_kms_client(app_ctx, &credentials, &client);
+
+    /* Generate data key with KMS. */
+    struct aws_byte_buf plaintext;
+    struct aws_byte_buf ciphertext;
+    rc = aws_kms_generate_data_key_blocking(client, app_ctx->key_id, app_ctx->key_spec, &plaintext, &ciphertext);
+    fail_on(rc != AWS_OP_SUCCESS, "Could not generate data key");
```

Comment:
Is there a way to report the underlying error from KMS here? I'm getting this error (or the one on line 443), and I can't tell what the problem is.

Reply:
At the moment I don't believe that the underlying C SDK is returning this data. While I agree that it would be a valuable enhancement, I don't think it should block merging this PR.

That said, it's possible that some additional information such as error codes are getting printed to console which would be visible in debug mode. I'd have to check. Additionally CloudTrail logs may provide some insight as well.

Reply:
Yeah, we plan at some point to pass KMS error logs to the caller since we would also like to improve the ACM for Nitro Enclaves user applications debugging in production.